stat/stat.go

```diff
@@ -24,6 +26,272 @@ const (
+		var pCDFAtX1 float64
+		for j := 0; j < len(pPairs); j++ {
+			if pPairs[j].value > x1 {
+				break
+			}
+			pCDFAtX1 = pCDF[j]
+		}
+
+		var qCDFAtX1 float64
+		for j := 0; j < len(qPairs); j++ {
+			if qPairs[j].value > x1 {
+				break
+			}
+			qCDFAtX1 = qCDF[j]
+		}
+
+		distance += float64(math.Abs(pCDFAtX1-qCDFAtX1) * (x2 - x1))
+	}
+
+	return distance
+}
+
+// WassersteinDistanceND computes the Wasserstein distance (Earth Mover's Distance)
+// between two n-dimensional distributions p and q with optional weights.
+//
+// Parameters:
+// p and q are matrices where each row represents a point with coordinates as columns.
+// pWeights and qWeights are optional weights for each point. If nil, uniform weights are used.
+//
+// This implementation uses linear programming to solve the optimal transport problem.
+func WassersteinDistanceND(p, q mat.Matrix, pWeights, qWeights []float64) (float64, error) {
+	pRows, pCols := p.Dims()
+	qRows, qCols := q.Dims()
+
+	if pCols == 0 || qCols == 0 || pRows == 0 || qRows == 0 {
+		return math.NaN(), nil
+	}
+
+	if (pWeights != nil && pRows != len(pWeights)) || (qWeights != nil && qRows != len(qWeights)) {
+		panic("stat: input distributions and their weights must have same length")
+	}
+
+	if pCols != qCols {
+		panic("stat: point dimensions must match between distributions")
+	}
+
+	// Special case for single-point distributions
+	if pRows == 1 && qRows == 1 {
+		var sumSquares float64
+		for j := 0; j < pCols; j++ {
+			diff := p.At(0, j) - q.At(0, j)
+			sumSquares += diff * diff
+		}
+		return math.Sqrt(sumSquares), nil
+	}
+
+	// Handle identical distributions case.
+	if pRows == qRows {
+		identical := true
+		for i := 0; i < pRows; i++ {
+			for j := 0; j < pCols; j++ {
+				if p.At(i, j) != q.At(i, j) {
+					identical = false
+					break
+				}
+			}
+		}
+		if identical {
+			if floats.Equal(pWeights, qWeights) {
+				return 0, nil
+			}
+		}
+	}
+
+	// Use uniform weights if not provided.
+	if pWeights == nil {
+		pWeights = make([]float64, pRows)
+		floats.AddConst(1/float64(pRows), pWeights)
+	} else {
+		normalizeWeights(pWeights)
+	}
+
+	if qWeights == nil {
+		qWeights = make([]float64, qRows)
+		floats.AddConst(1/float64(qRows), qWeights)
+	} else {
+		normalizeWeights(qWeights)
+	}
+
+	// Create cost matrix using Euclidean distance.
+	cost := mat.NewDense(pRows, qRows, nil)
+
+	for i := 0; i < pRows; i++ {
+		for j := 0; j < qRows; j++ {
+			// Calculate Euclidean distance between points
+			var sumSquares float64
+			for k := 0; k < pCols; k++ {
+				diff := p.At(i, k) - q.At(j, k)
+				sumSquares += diff * diff
+			}
+			dist := math.Sqrt(sumSquares)
+			cost.Set(i, j, dist)
+		}
+	}
+
+	return solveOptimalTransportLP(cost, pWeights, qWeights)
+}
+
+// normalizeWeights checks and normalizes the provided weight array if needed.
+func normalizeWeights(weights []float64) {
+	total := floats.Sum(weights)
+
+	if total <= 0 {
+		panic("stat: weights must sum to positive values")
+	}
```

Comment:
This check is not correct; it allows negative weights to exist so long as the total is positive.